pkg/api/defaults-custom-cloud-profile.go

```diff
@@ -3,34 +3,110 @@
+		if p.CustomCloudProfile.Environment == nil {
+			p.CustomCloudProfile.Environment = &azure.Environment{}
 		}
 
-		//Set default value for ResourceManagerVMDNSSuffix
-		azureStackCloudSpec.EndpointConfig.ResourceManagerVMDNSSuffix = p.CustomCloudProfile.Environment.ResourceManagerVMDNSSuffix
+		env := p.CustomCloudProfile.Environment
+		if env.Name == "" || env.ResourceManagerEndpoint == "" || env.ServiceManagementEndpoint == "" || env.ActiveDirectoryEndpoint == "" || env.GraphEndpoint == "" || env.ResourceManagerVMDNSSuffix == "" {
+			env.Name = AzureStackCloud
+			azsFQDNSuffix := strings.Replace(p.CustomCloudProfile.PortalURL, fmt.Sprintf("https://portal.%s.", cs.Location), "", -1)
+			azsFQDNSuffix = strings.TrimSuffix(azsFQDNSuffix, "/")
+			env.ResourceManagerEndpoint = fmt.Sprintf("https://management.%s.%s/", cs.Location, azsFQDNSuffix)
+			metadataURL := fmt.Sprintf("%s/metadata/endpoints?api-version=1.0", strings.TrimSuffix(env.ResourceManagerEndpoint, "/"))
 
+			// Retrieve the metadata
+			httpClient := &http.Client{
+				Timeout: 30 * time.Second,
+			}
+			endpointsresp, err := httpClient.Get(metadataURL)
+			if err != nil || endpointsresp.StatusCode != 200 {
+				return fmt.Errorf("%s . apimodel invalid: failed to retrieve Azure Stack endpoints from %s", err, metadataURL)
+			}
+
+			body, err := ioutil.ReadAll(endpointsresp.Body)
+			if err != nil {
+				return fmt.Errorf("%s . apimodel invalid: failed to read the response from %s", err, metadataURL)
+			}
+
+			endpoints := AzureStackMetadataEndpoints{}
+			err = json.Unmarshal(body, &endpoints)
+			if err != nil {
+				return fmt.Errorf("%s . apimodel invalid: failed to parse the response from %s", err, metadataURL)
+			}
+
+			if endpoints.GraphEndpoint == "" || endpoints.Authentication == nil || endpoints.Authentication.LoginEndpoint == "" || len(endpoints.Authentication.Audiences) == 0 || endpoints.Authentication.Audiences[0] == "" {
+				return fmt.Errorf("%s . apimodel invalid: invalid response from %s", err, metadataURL)
+			}
+
+			env.GraphEndpoint = endpoints.GraphEndpoint
+			env.ServiceManagementEndpoint = endpoints.Authentication.Audiences[0]
+			env.GalleryEndpoint = endpoints.GalleryEndpoint
+			env.ActiveDirectoryEndpoint = endpoints.Authentication.LoginEndpoint
+			if p.CustomCloudProfile.IdentitySystem == ADFSIdentitySystem {
+				env.ActiveDirectoryEndpoint = strings.TrimSuffix(env.ActiveDirectoryEndpoint, "/")
+				env.ActiveDirectoryEndpoint = strings.TrimSuffix(env.ActiveDirectoryEndpoint, "adfs")
+			}
+
+			env.ManagementPortalURL = endpoints.PortalEndpoint
+			env.ResourceManagerVMDNSSuffix = fmt.Sprintf("cloudapp.%s", azsFQDNSuffix)
+			env.StorageEndpointSuffix = fmt.Sprintf("%s.%s", cs.Location, azsFQDNSuffix)
+			env.KeyVaultDNSSuffix = fmt.Sprintf("vault.%s.%s", cs.Location, azsFQDNSuffix)
+		}
+	}
+	return nil
+}
+
+// SetCustomAzureCloudSpec sets the cloud spec for Azure Stack .
+func (p *Properties) SetCustomAzureCloudSpec() {
```

Comment:
Do we think this function will grow to include other types of customizations? If not (or not right away), maybe it would be clearer to include AzureStack in the name, like `SetAzureStackCloudSpec`.

Reply:
I will make the change.